rest/oidc_api_test.go

```diff
@@ -1,139 +1,791 @@
+			if assertWantErrType(t, response, tc.wantErrType, grantTypeRefreshToken) {
+				return
+			}
+			require.Equal(t, http.StatusOK, response.StatusCode)
+
+			// Validate received token refresh response.
+			require.NoError(t, err, json.NewDecoder(response.Body).Decode(&receivedToken))
+			require.NoError(t, response.Body.Close(), "Error closing response body")
+			wantTokenResponse = mockAuthServer.options.tokenResponse
+			assert.NotEmpty(t, receivedToken.SessionID, "session_id doesn't exist")
+			assert.Equal(t, "foo_noah", receivedToken.Username, "name mismatch")
+			assert.Equal(t, wantTokenResponse.IDToken, receivedToken.IDToken, "id_token mismatch")
+			assert.Equal(t, wantTokenResponse.RefreshToken, receivedToken.RefreshToken, "refresh_token mismatch")
+			if tc.providers["foo"].IncludeAccessToken {
+				assert.Equal(t, wantTokenResponse.AccessToken, receivedToken.AccessToken, "access_token mismatch")
+				assert.Equal(t, wantTokenResponse.TokenType, receivedToken.TokenType, "token_type mismatch")
+				assert.True(t, wantTokenResponse.Expires >= receivedToken.Expires, "expires_in mismatch")
+			}
+			// Query db endpoint with Bearer token
+			request, err = http.NewRequest(http.MethodGet, dbEndpoint, nil)
+			require.NoError(t, err, "Error creating new request")
+			request.Header.Add("Authorization", receivedToken.IDToken)
+			response, err = http.DefaultClient.Do(request)
+			require.NoError(t, err, "Error sending request with bearer token")
+			require.Equal(t, http.StatusOK, response.StatusCode)
+			require.NoError(t, err, json.NewDecoder(response.Body).Decode(&responseBody))
+			require.NoError(t, response.Body.Close(), "Error closing response body")
+			assert.Equal(t, restTester.DatabaseConfig.Name, responseBody["db_name"])
+		})
+	}
+}
+
+// The assertHttpResponse asserts the statusCode and bodyText against statusCode and string
+// representation of the body bytes from the HTTP response.
+func assertHttpResponse(t *testing.T, response *http.Response, statusCode int, bodyText string) {
+	bodyBytes, err := ioutil.ReadAll(response.Body)
+	require.NoError(t, err, "error reading response body")
+	assert.Contains(t, string(bodyBytes), bodyText)
+	assert.Equal(t, statusCode, response.StatusCode)
+	require.NoError(t, response.Body.Close(), "error closing response body")
+}
+
+// refreshProviderConfig updates the issuer URL and sets the discovery endpoint
+// in provider configuration.
+func refreshProviderConfig(providers auth.OIDCProviderMap, issuer string) {
+	for name, provider := range providers {
+		provider.Issuer = issuer + "/" + name
+		provider.DiscoveryURI = provider.Issuer + auth.OIDCDiscoveryConfigPath
+	}
+}
+
+// assertWantErrType assert the given wantErrType on the response object. Return true if the wantErrType matches
+// with the error in response and false otherwise.
+func assertWantErrType(t *testing.T, response *http.Response, wantErrType wantErrType, grantType grantType) (exit bool) {
+	switch wantErrType {
```

Comment:
wantErrType struct
  expectedErrorCode
  expectedMessage
assertHttpResponse(t, response, wantErrType.code, wantErrType.Message)

Reply:
This change has been committed and ready for review.